### PR TITLE
[ML] move to using new BERT style tokenizer and analyzer instead of custom tokenization class

### PR DESCRIPTION
**Pull request by benwtrent**
This changes the internal tokenization logic to utilize Lucene style token filters and analyzers. 

These filters and analyzers are not generally available yet via the ML plugin and are only used internally.

Micro benchmarks show a marginal improvement on average on throughput
```
BertEncodingBenchmark.luceneStyle  avgt   15  565.577 ±  3.527  ms/op
BertEncodingBenchmark.original     avgt   15  714.827 ± 15.393  ms/op
```

I think there is a ton of room for optimizing these tokenizers further. But, to keep readability for the initial transition, filters are created separately, and their implementations are rather naive.

### CODE REVIEW

**Comment by benwtrent** [comment]
@elasticmachine update branch

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/WordPieceTokenFilter.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+    public void reset() throws IOException {
+        super.reset();
+        tokens.clear();
+        tokenizedValues.clear();
+        current = null;
+    }
+
+    @Override
+    public boolean incrementToken() throws IOException {
+        // TODO seems like this + lowercase + tokenize cjk + wordpiece could all be the same thing....
```
Comment (droberts195) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/WordPieceTokenFilter.java:94]: Yes, I agree.  The different filters are not being composed in the analyzer config, they're being composed within this token filter.  So they might as well all be part of this same token filter, which will mean fewer intermediate data structures.  The only reason for keeping them completely separate is if we wanted the building blocks to be reusable generally within Elasticsearch.  But Elasticsearch must already have similar functionality available.  Therefore, it's better that eventually we publicly expose "word_piece" and it has all these options built in.  Then there's no need to have several classes extending `TokenFilter` that are not registered as generally available token filters.

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/WordPieceTokenFilter.java:94]: @droberts195 I do think have the "basic" tokenizer separate from wordpiece makes sense as I think it is used by other tokenization schemes. 

But I can totally collapse the other things into `basic` to make it just two filters

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/ControlCharFilter.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+                int start = pos;
+                while (start < cnt) {
+                    int category = Character.getType(temp[start]);
+                    if (category < Character.CONTROL || category > Character.SURROGATE) {
+                        break;
+                    }
+                    start++;
+                }
+                if (start > pos) {
+                    addOffCorrectMap(pos, start);
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/ControlCharFilter.java:64]: The java doc for `addOffCorrectMap` notes

> Assumption: the offset given with each successive call to this method will not be smaller than the offset given at the previous invocation.

If the `for` loop is executed multiple times the offset `pos` could be smaller as it is reset to 0. Should the offset include the total number of chars read in previous iterations?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/CharSeqTokenTrieNode.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+import org.apache.lucene.analysis.CharArrayMap;
+import org.elasticsearch.core.CheckedFunction;
+import org.elasticsearch.core.Nullable;
+
+import java.io.IOException;
+import java.util.Collection;
+import java.util.List;
+import java.util.Objects;
+
+public class CharSeqTokenTrieNode {
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/CharSeqTokenTrieNode.java:19]: This doesn't look like it's a move from `TokenTrieNode.java`. Do we still have that file? Should it be deleted?

Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/CharSeqTokenTrieNode.java:19]: `TokenTrieNode` should be deleted. git might not have treated it like a move. I will look.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/ControlCharFilter.java (hunk 2 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/ControlCharFilter.java:64]: Good catch! None of my tests test a 1024+ length string.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/DelimitedToken.java [hunk 4]
```diff
@@ -8,65 +8,43 @@
...
     }
 
-    public String getToken() {
-        return token;
+    public CharSequence charSequence() {
+        return charSequence;
     }
 
-    @Override
-    public boolean equals(Object o) {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/DelimitedToken.java:56]: I can put them back, it was initially a record, but some restrictions disallowed that (sub-classing).

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 5]
```diff
@@ -0,0 +1,272 @@
...
+            DelimitedToken token = tokens.removeFirst();
+            termAtt.setEmpty().append(token.charSequence());
+            offsetAtt.setOffset(token.startOffset(), token.endOffset());
+            return true;
+        }
+        return false;
+    }
+
+    void stripAccent() {
+        if (normalizer.quickCheck(termAtt) != Normalizer.YES) {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:221]: Yeah, this is a bug, will fix. Basically, we should quick check before normalizing, and only normalize if required. Then continue to strip accents.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/ControlCharFilterFactory.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+package org.elasticsearch.xpack.ml.inference.nlp.tokenizers;
+
+import org.elasticsearch.common.settings.Settings;
+import org.elasticsearch.env.Environment;
+import org.elasticsearch.index.IndexSettings;
+import org.elasticsearch.index.analysis.AbstractCharFilterFactory;
+
+import java.io.Reader;
+
+public class ControlCharFilterFactory extends AbstractCharFilterFactory {
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/ControlCharFilterFactory.java]: nope, removed.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+        if (input.incrementToken()) {
+            if (isStripAccents) {
+                stripAccent();
+            }
+            if (neverSplitSet.contains(termAtt)) {
+                return true;
+            }
+            int startOffset = offsetAtt.startOffset();
+            // split punctuation and maybe cjk chars!!!
+            LinkedList<DelimitedToken> splits = new LinkedList<>();
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:134]: This method is too long. Could we extract this bit all the way to line 169 into a `splitPunctuation` sort of method?

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+                        lastCharSplit + startOffset,
+                        offsetAtt.endOffset()
+                    )
+                );
+            }
+            // There is nothing to merge, nothing to store, simply return
+            if (splits.size() == 1) {
+                return true;
+            }
+            List<DelimitedToken> matchingTokens = new ArrayList<>();
```
Comment (dimitris-athanasiou) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:174]: This part should also be possible to be extracted to a method

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 9]
```diff
@@ -0,0 +1,273 @@
...
+    private final LinkedList<DelimitedToken> tokens;
+    private final boolean isStripAccents;
+    private final CharArraySet neverSplitSet;
+    private final Normalizer2 normalizer;
+    private final StringBuilder accentBuffer = new StringBuilder();
+    private final IntPredicate splitOn;
+
+    private State current;
+
+    public static BasicTokenFilter buildFromSettings(
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:47]: Sure, I can rename it. Eventually, there will be a settings object (thats how analyzers get their unique settings per index).

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java (hunk 7 above)
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:134]: for sure

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilterTests.java [hunk 10]
```diff
@@ -0,0 +1,134 @@
...
+
+    public void testSplitCJK() throws Exception {
+        Analyzer analyzer = basicAnalyzerFromSettings(true, false, List.of("[UNK]"));
+        assertAnalyzesToNoCharFilter(analyzer, "hello ah\u535A\u63A8zz", new String[] { "hello", "ah", "\u535A", "\u63A8", "zz" });
+        assertAnalyzesToNoCharFilter(analyzer, "hello world", new String[] { "hello", "world" });
+    }
+
+    public void testStripAccents() throws Exception {
+        Analyzer analyzer = basicAnalyzerFromSettings(true, true, List.of("[UNK]"));
+        assertAnalyzesToNoCharFilter(analyzer, "HäLLo how are you", new String[] { "HaLLo", "how", "are", "you" });
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilterTests.java:69]: nit: add NFD version of the test, e.g. `h\u0061\u0308llo ...`

File: x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilterTests.java [hunk 11]
```diff
@@ -0,0 +1,134 @@
...
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK]?", new String[] { "Hello", "[UNK]", "?" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello [UNK]!!", new String[] { "Hello", "[UNK]", "!", "!" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello-[UNK]", new String[] { "Hello", "-", "[UNK]" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello~[UNK][UNK]", new String[] { "Hello", "~", "[UNK]", "[UNK]" });
+        assertAnalyzesToNoCharFilter(analyzer, "Hello-[unk]", new String[] { "Hello", "-", "[", "unk", "]" });
+    }
+
+    public void testSplitCJK() throws Exception {
+        Analyzer analyzer = basicAnalyzerFromSettings(true, false, List.of("[UNK]"));
+        assertAnalyzesToNoCharFilter(analyzer, "hello ah\u535A\u63A8zz", new String[] { "hello", "ah", "\u535A", "\u63A8", "zz" });
```
Comment (hendrikmuhs) [inline x-pack/plugin/ml/src/test/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilterTests.java:63]: nit: what happens for surrogate pairs? are they getting split or are they 1 token?

**Review by hendrikmuhs — COMMENTED** [review]
cool to migrate to lucene analyzers, added some comments

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 12]
```diff
@@ -0,0 +1,272 @@
...
+            index++;
+        }
+        if (badIndices.isEmpty()) {
+            return;
+        }
+        for (int i = 0; i < badIndices.size(); i++) {
+            int badIndex = badIndices.get(i);
+            int count = charCount.get(i);
+            for (int j = 0; j < count && badIndex < accentBuffer.length(); j++) {
+                accentBuffer.deleteCharAt(badIndex);
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:167]: Pretty sure string buffers are safe to delete in the middle of.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 13]
```diff
@@ -0,0 +1,272 @@
...
+    private final LinkedList<DelimitedToken> tokens;
+    private final boolean isStripAccents;
+    private final CharArraySet neverSplitSet;
+    private final Normalizer2 normalizer;
+    private final StringBuilder accentBuffer = new StringBuilder();
+    private final IntPredicate splitOn;
+
+    private State current;
+
+    public static BasicTokenFilter build(boolean isTokenizeCjkChars, boolean isStripAccents, List<String> neverSplit, TokenStream input)
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:47]: @hendrikmuhs this is how it was before.

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java [hunk 14]
```diff
@@ -0,0 +1,272 @@
...
+            return;
+        }
+        for (int i = 0; i < badIndices.size(); i++) {
+            int badIndex = badIndices.get(i);
+            int count = charCount.get(i);
+            for (int j = 0; j < count && badIndex < accentBuffer.length(); j++) {
+                accentBuffer.deleteCharAt(badIndex);
+            }
+        }
+        termAtt.setEmpty().append(accentBuffer);
```
[the remark this answers is not in the document]
Comment (benwtrent) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:170]: Again, I haven't a clue :), but this is effectively what it was before

**Comment by benwtrent** [comment]
All your comments @hendrikmuhs seem to be about how tokenization is done technically. I am simply moving the the tokenization into a lucene format. Any improvements and changes to the tokenization should be done separately

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (hendrikmuhs) [reply x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/nlp/tokenizers/BasicTokenFilter.java:167]: yes, my comment was about runtime not whether it is safe or not.

**Comment by hendrikmuhs** [comment]
@benwtrent sure, don't feel offended. Comments are all marked as _nits_ - treat them as you prefer. It wasn't obvious to me what has been moved and what is new.

**Comment by benwtrent** [comment]
@hendrikmuhs no offense taken! I think they are valid suggestions and things we should consider. We should create a GH issue to capture these ideas and do a formal investigation into their impacts. 

Could you do that?

**Comment by benwtrent** [comment]
@elasticmachine update branch

**Comment by benwtrent** [comment]
run elasticsearch-ci/part-1

**Review by dimitris-athanasiou — APPROVED** [review]
LGTM

**Review by droberts195 — APPROVED** [review]
LGTM

If this is doing what the original tokenizer outside the Lucene framework was doing then I'm happy to merge this and continue to iterate in future PRs.

This isn't yet exposed to end users as a generic analyzer and token filter that they can use in any context.  At some point in the future we need to make them available in that way, but first let's use them internally ourselves for a release or two to smoke out any problems before the public interface is hard to change.

**Pull request merged**